File: VeteranGallery.Domain/Entities/Veteran.cs
using System.Text.Json.Serialization;
using VeteranGallery.Domain.Enums;

namespace VeteranGallery.Domain.Entities;

[JsonPolymorphic(TypeDiscriminatorPropertyName = "$type")]
[JsonDerivedType(typeof(Infantryman), typeDiscriminator: "infantry")]
[JsonDerivedType(typeof(Artillery), typeDiscriminator: "artillery")]
[JsonDerivedType(typeof(TankCrewman), typeDiscriminator: "tank_crew")]
[JsonDerivedType(typeof(Pilot), typeDiscriminator: "pilot")]
[JsonDerivedType(typeof(AirDefenseOperator), typeDiscriminator: "air_defense")]
[JsonDerivedType(typeof(FlightNavigator), typeDiscriminator: "navigator")]
[JsonDerivedType(typeof(DroneOperator), typeDiscriminator: "drone_op")]
[JsonDerivedType(typeof(Paratrooper), typeDiscriminator: "paratrooper")]
[JsonDerivedType(typeof(AirAssaultSapper), typeDiscriminator: "assault_sapper")]
[JsonDerivedType(typeof(NavySailor), typeDiscriminator: "navy")]
[JsonDerivedType(typeof(CombatDiver), typeDiscriminator: "combat_diver")]
[JsonDerivedType(typeof(NavalArtillerist), typeDiscriminator: "naval_artillery")]
[JsonDerivedType(typeof(SpecialForcesSoldier), typeDiscriminator: "special_ops")]
[JsonDerivedType(typeof(Sniper), typeDiscriminator: "sniper")]
[JsonDerivedType(typeof(SpecialOpsIntelligence), typeDiscriminator: "spec_intel")]
public abstract class Veteran
{
    private Guid _id = Guid.NewGuid();
    public Guid Id { get => _id; init => _id = value; }

    public bool IsDeleted { get; set; } = false;
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public string FullName { get; set; } = string.Empty;
    public MilitaryRank Rank { get; set; }
    public string UnitName { get; set; } = string.Empty;
    public string Story { get; set; } = string.Empty;
    public MilitaryBranch Branch { get; set; }
    public string? PhotoUrl { get; set; }

    public virtual string BranchDisplayName => "Armed Forces";

    public abstract string GetSpecializedDescription();

    public void RegenerateId() => _id = Guid.NewGuid();

    public void Validate()
    {
        if (string.IsNullOrWhiteSpace(FullName))
            throw new ArgumentException("Veteran's full name cannot be empty.", nameof(FullName));
        if (string.IsNullOrWhiteSpace(UnitName))
            throw new ArgumentException("Unit name cannot be empty.", nameof(UnitName));
        if (string.IsNullOrWhiteSpace(Story))
            throw new ArgumentException("Veteran's story cannot be empty.", nameof(Story));
    }

    public virtual string GetSearchableText() => string.Empty;

    public bool IsSeniorOfficer() => (int)Rank >= (int)MilitaryRank.BrigadierGeneralOrCommodore;

    public string GetRankDisplayName() => Rank switch
    {
        MilitaryRank.Recruit                                             => "Recruit",
        MilitaryRank.SoldierOrSailor                                     => "Soldier / Sailor",
        MilitaryRank.SeniorSoldierOrSeniorSailor                         => "Senior Soldier / Senior Sailor",
        MilitaryRank.JuniorSergeantOrPettyOfficer2ndClass                => "Junior Sergeant / Petty Officer 2nd Class",
        MilitaryRank.SergeantOrPettyOfficer1stClass                      => "Sergeant / Petty Officer 1st Class",
        MilitaryRank.SeniorSergeantOrChiefPettyOfficer                   => "Senior Sergeant / Chief Petty Officer",
        MilitaryRank.ChiefSergeantOrChiefShipPettyOfficer                => "Chief Sergeant / Chief Ship Petty Officer",
        MilitaryRank.StaffSergeantOrStaffWarrantOfficer                  => "Staff Sergeant / Staff Warrant Officer",
        MilitaryRank.MasterSergeantOrMasterWarrantOfficer                => "Master Sergeant / Master Warrant Officer",
        MilitaryRank.SeniorMasterSergeantOrSeniorMasterWarrantOfficer    => "Senior Master Sergeant / Senior Master Warrant Officer",
        MilitaryRank.ChiefMasterSergeantOrChiefMasterWarrantOfficer      => "Chief Master Sergeant / Chief Master Warrant Officer",
        MilitaryRank.JuniorLieutenant                                    => "Junior Lieutenant",
        MilitaryRank.Lieutenant                                          => "Lieutenant",
        MilitaryRank.SeniorLieutenant                                    => "Senior Lieutenant",
        MilitaryRank.CaptainOrCaptainLieutenant                          => "Captain / Captain-Lieutenant",
        MilitaryRank.MajorOrCaptain3rdRank                               => "Major / Captain 3rd Rank",
        MilitaryRank.LieutenantColonelOrCaptain2ndRank                   => "Lieutenant Colonel / Captain 2nd Rank",
        MilitaryRank.ColonelOrCaptain1stRank                             => "Colonel / Captain 1st Rank",
        MilitaryRank.BrigadierGeneralOrCommodore                         => "Brigadier General / Commodore",
        MilitaryRank.MajorGeneralOrRearAdmiral                           => "Major General / Rear Admiral",
        MilitaryRank.LieutenantGeneralOrViceAdmiral                      => "Lieutenant General / Vice Admiral",
        MilitaryRank.GeneralOrAdmiral                                    => "General / Admiral",
        _                                                                => Rank.ToString()
    };
}

File: veterangallery.client/src/types/veteran.ts
export const MilitaryBranch = {
    LandForces: 1,
    AirForce: 2,
    AirAssault: 3,
    Navy: 4
} as const;

export type MilitaryBranch = typeof MilitaryBranch[keyof typeof MilitaryBranch];

export const MilitaryRank = {
    Recruit: 1, 
    SoldierOrSailor: 2, 
    SeniorSoldierOrSeniorSailor: 3,
    JuniorSergeantOrPettyOfficer2ndClass: 4, 
    SergeantOrPettyOfficer1stClass: 5,
    SeniorSergeantOrChiefPettyOfficer: 6, 
    ChiefSergeantOrChiefShipPettyOfficer: 7,
    StaffSergeantOrStaffWarrantOfficer: 8, 
    MasterSergeantOrMasterWarrantOfficer: 9,
    SeniorMasterSergeantOrSeniorMasterWarrantOfficer: 10, 
    ChiefMasterSergeantOrChiefMasterWarrantOfficer: 11,
    JuniorLieutenant: 12, 
    Lieutenant: 13, 
    SeniorLieutenant: 14, 
    CaptainOrCaptainLieutenant: 15,
    MajorOrCaptain3rdRank: 16, 
    LieutenantColonelOrCaptain2ndRank: 17, 
    ColonelOrCaptain1stRank: 18,
    BrigadierGeneralOrCommodore: 19, 
    MajorGeneralOrRearAdmiral: 20, 
    LieutenantGeneralOrViceAdmiral: 21, 
    GeneralOrAdmiral: 22
} as const;

export type MilitaryRank = typeof MilitaryRank[keyof typeof MilitaryRank];

export interface Veteran {
    id: string;
    fullName: string;
    rank: MilitaryRank;
    unitName: string;
    story: string;
    branch: MilitaryBranch;
    photoUrl?: string;
    $type: string;
}

export interface Pilot extends Veteran {
    $type: 'pilot';
    vehicleModel: string;
    experienceValue: number;
}

export interface Infantryman extends Veteran {
    $type: 'infantry';
    specialization: string;
}

export interface DroneOperator extends Veteran {
    $type: 'drone_op';
    vehicleModel: string;
    experienceValue: number;
}

export const getRankDisplayName = (rank: MilitaryRank, branch: MilitaryBranch): string => {
    const isNavy = branch === MilitaryBranch.Navy;
    
    switch (rank) {
        case MilitaryRank.Recruit: return 'Recruit';
        case MilitaryRank.SoldierOrSailor: return isNavy ? 'Seaman' : 'Soldier';
        case MilitaryRank.SeniorSoldierOrSeniorSailor: return isNavy ? 'Senior Seaman' : 'Senior Soldier';
        
        case MilitaryRank.JuniorSergeantOrPettyOfficer2ndClass: return isNavy ? 'Petty Officer 2nd Class' : 'Junior Sergeant';
        case MilitaryRank.SergeantOrPettyOfficer1stClass: return isNavy ? 'Petty Officer 1st Class' : 'Sergeant';
        case MilitaryRank.SeniorSergeantOrChiefPettyOfficer: return isNavy ? 'Chief Petty Officer' : 'Senior Sergeant';
        case MilitaryRank.ChiefSergeantOrChiefShipPettyOfficer: return isNavy ? 'Chief Ship Petty Officer' : 'Chief Sergeant';
        
        case MilitaryRank.StaffSergeantOrStaffWarrantOfficer: return isNavy ? 'Staff Warrant Officer' : 'Staff Sergeant';
        case MilitaryRank.MasterSergeantOrMasterWarrantOfficer: return isNavy ? 'Master Warrant Officer' : 'Master Sergeant';
        case MilitaryRank.SeniorMasterSergeantOrSeniorMasterWarrantOfficer: return isNavy ? 'Senior Master Warrant Officer' : 'Senior Master Sergeant';
        case MilitaryRank.ChiefMasterSergeantOrChiefMasterWarrantOfficer: return isNavy ? 'Chief Master Warrant Officer' : 'Chief Master Sergeant';

        case MilitaryRank.JuniorLieutenant: return 'Junior Lieutenant';
        case MilitaryRank.Lieutenant: return 'Lieutenant';
        case MilitaryRank.SeniorLieutenant: return 'Senior Lieutenant';
        case MilitaryRank.CaptainOrCaptainLieutenant: return isNavy ? 'Captain-Lieutenant' : 'Captain';
        
        case MilitaryRank.MajorOrCaptain3rdRank: return isNavy ? 'Captain 3rd Rank' : 'Major';
        case MilitaryRank.LieutenantColonelOrCaptain2ndRank: return isNavy ? 'Captain 2nd Rank' : 'Lieutenant Colonel';
        case MilitaryRank.ColonelOrCaptain1stRank: return isNavy ? 'Captain 1st Rank' : 'Colonel';
        
        case MilitaryRank.BrigadierGeneralOrCommodore: return isNavy ? 'Commodore' : 'Brigadier General';
        case MilitaryRank.MajorGeneralOrRearAdmiral: return isNavy ? 'Rear Admiral' : 'Major General';
        case MilitaryRank.LieutenantGeneralOrViceAdmiral: return isNavy ? 'Vice Admiral' : 'Lieutenant General';
        case MilitaryRank.GeneralOrAdmiral: return isNavy ? 'Admiral' : 'General';
        
        default: return 'Unknown Rank';
    }
};